File: JAVA/src/main/scala/engine/math/Quaternion.scala
package engine.math

/**
  * Created by Mnenmenth Alkaborin
  * Please refer to LICENSE file if included
  * for licensing information
  * https://github.com/Mnenmenth
  */
class Quaternion(x1: Float, y1: Float, z1: Float, w1: Float) {

  var x: Float = x1
  var y: Float = y1
  var z: Float = z1
  var w: Float = w1

  def length(): Float = Math.sqrt(x * x + y * y + z * z + w * w).toFloat
  def conjugate(): Quaternion = new Quaternion(-x, -y, -z, w)

  def normalize(): Quaternion = {
    val l: Float = length
    x /= l
    y /= l
    z /= l
    w /= l
    this
  }

  def mult(q: Quaternion): Quaternion = {
    val w2: Float = w * q.w - x * q.x - y * q.y - z * q.z
    val x2: Float = x * q.w + w * q.x + y * q.z - z * q.y
    val y2: Float = y * q.w + w * q.y + z * q.x - x * q.z
    val z2: Float = z * q.w + w * q.z + x * q.y - y * q.z
    new Quaternion(x2, y2, z2, w2)
  }

  def mult(v: Vector3f): Quaternion = {
    val w2: Float = -x * v.x - y * v.y - z * v.z
    val x2: Float = w * v.x + y * v.z - z * v.y
    val y2: Float = w * v.y + z * v.x - x * v.z
    val z2: Float = w * v.z + x * v.y - y * v.x
    new Quaternion(x2, y2, z2, w2)
  }

}

File: CPP/src/engine/math/Quaternion.cpp
//
// Created by Razim on 1/9/2016.
//

#include "Quaternion.h"
#include <math.h>

Quaternion::Quaternion(float x, float y, float z, float w) {
    Quaternion::x = x;
    Quaternion::y = y;
    Quaternion::z = z;
    Quaternion::w = w;
}

float Quaternion::length() {
    return (float) sqrt(pow(Quaternion::x, 2) + pow(Quaternion::y, 2) + pow(Quaternion::z, 2) + pow(Quaternion::w, 2));
}

Quaternion Quaternion::normalize() {
    float length = Quaternion::length();
    Quaternion::x /= length;
    Quaternion::y /= length;
    Quaternion::z /= length;
    Quaternion::w /= length;
    return *this;
}

Quaternion Quaternion::conjugate() {
    return Quaternion(-Quaternion::x, -Quaternion::y, -Quaternion::z, w);
}

Quaternion Quaternion::mult(Quaternion q) {
    float w = Quaternion::w * q.getW() - Quaternion::x * q.getX() - Quaternion::y * q.getY() - Quaternion::z * q.getZ();
    float x = Quaternion::x * q.getW() + Quaternion::w * q.getX() + Quaternion::y * q.getZ() - Quaternion::z * q.getY();
    float y = Quaternion::y * q.getW() + Quaternion::w * q.getY() + Quaternion::z * q.getX() - Quaternion::x * q.getZ();
    float z = Quaternion::z * q.getW() + Quaternion::w * q.getZ() + Quaternion::x * q.getY() - Quaternion::y * q.getZ();
    return Quaternion(x, y, z, w);
}

Quaternion Quaternion::mult(Vector3f v) {
    float w = -Quaternion::x * v.getX() - Quaternion::y * v.getY() - Quaternion::z * v.getZ();
    float x = Quaternion::w * v.getX() + Quaternion::y * v.getZ() - Quaternion::z * v.getY();
    float y = Quaternion::w * v.getY() + Quaternion::z * v.getX() - Quaternion::x * v.getZ();
    float z = Quaternion::w * v.getZ() + Quaternion::x * v.getY() - Quaternion::y * v.getX();
    return Quaternion(x, y, z, w);
}

float Quaternion::getX() {
    return Quaternion::x;
}

void Quaternion::setX(float x) {
    Quaternion::x = x;
}

float Quaternion::getY() {
    return Quaternion::y;
}

void Quaternion::setY(float y) {
    Quaternion::y = y;
}

float Quaternion::getZ() {
    return Quaternion::z;
}

void Quaternion::setZ(float z) {
    Quaternion::z = z;
}

float Quaternion::getW() {
    return Quaternion::w;
}

void Quaternion::setW(float w) {
    Quaternion::w = w;
}